python/lsst/daf/butler/tests/cliOptionTestBase.py

```diff
@@ -44,37 +336,275 @@ def run_command(self, cmd, args):
+    and handles that argument correctly.
+    """
+
+    def test_required_missing(self):
+        if self.isArgument:
+            expected = f'Missing argument "{self.optionName.upper()}"'
+        else:
+            if self.shortOptionFlag:
+                expected = f'Missing option "{self.shortOptionFlag}" / "{self.optionFlag}"'
+            else:
+                expected = f'Missing option "--{self.optionName}"'
+
+        self.run_test(CliFactory.noOp(self, parameterKwargs=dict(required=True)),
+                      [],
+                      self.verifyMissing,
+                      expected)
+
+    def _test_forRequired_provided(self, required):
+        def doTest(self):
+            helper = CliFactory.mocked(self,
+                                       expectedKwargs={self.optionKey: self.expectedVal},
+                                       parameterKwargs=dict(required=required))
+            self.run_test(helper.cli,
+                          self.makeInputs(self.optionFlag, self.optionValue),
+                          self.verifyCalledWith,
+                          helper)
+
+        if type(self.valueType) == click.Path:
+            OptPathTypeTest.runForPathType(self, doTest)
+        else:
+            doTest(self)
+
+    def test_required_provided(self):
+        self._test_forRequired_provided(required=True)
+
+    def test_required_notRequiredProvided(self):
+        self._test_forRequired_provided(required=False)
+
+    def test_required_notRequiredDefaultValue(self):
+        """Verify that the expected default value is passed for a paramter when
+        it is not used on the command line."""
+        helper = CliFactory.mocked(self,
+                                   expectedKwargs={self.optionKey: self.expectedValDefault})
+        self.run_test(helper.cli,
+                      [],
+                      self.verifyCalledWith,
+                      helper)
+
+
+class OptPathTypeTest(OptTestBase):
+    """A mixin that tests options that have `type=click.Path`
+    """
+
+    @staticmethod
+    def runForPathType(testObj, testFunc):
+        """Function to execute the path type test, sets up directories and a
+        file as needed by the test.
+
+        Parameters
+        ----------
+        testObj : `OptTestBase` instance
+            The `OptTestBase` subclass that is running the test.
+        testFunc : A callable function that takes no args.
+            The function that executes the test.
+        """
+        with testObj.runner.isolated_filesystem():
+            if testObj.valueType.exists:
+                # If the file or dir is expected to exist, create it since it's
+                # it doesn't exist because we're running in a temporary
+                # directory.
+                if testObj.valueType.dir_okay:
+                    os.makedirs(testObj.optionValue)
+                elif testObj.valueType.file_okay:
+                    _ = open(testObj.optionValue)
```

Comment:
My guess is that you are creating a file here, should this be `open(..., "w")`? And I think it needs an explicit `close()` or `with open(...)`.